File: src/main/scala/com/vc/medium/L723.scala
package com.vc.medium

object L723 {
  def candyCrush(board: Array[Array[Int]]): Array[Array[Int]] = {
    val n = board.length
    val m = board(0).length
    import scala.collection.mutable
    val set = new mutable.HashSet[(Int, Int)]()
    var found = true
    while(found) {
      found = false
      (0 until n).foreach(i => {
        (0 until m).foreach(j => {
          val e = board(i)(j)
          if(e != 0) {
            if(i + 2 < n && board(i + 1)(j) == e && board(i + 2)(j) == e) {
              found = true
              var k = i
              while(k < n && board(k)(j) == e) {
                set += ((k, j))
                k += 1
              }
            }
            if(j + 2 < m && board(i)(j + 1) == e && board(i)(j + 2) == e) {
              found = true
              var l = j
              while(l < m && board(i)(l) == e) {
                set += ((i, l))
                l += 1
              }
            }
          }
        })
      })

      (0 until m).foreach(j => {
        var top = n - 1
        var bottom = n - 1
        while(top >= 0) {
          if(set.contains((top, j))) {
            board(top)(j) = 0
            top -= 1
          }
          else {
            board(bottom)(j) = board(top)(j)
            bottom -= 1
            top -= 1
          }
        }
        while(bottom >= 0) {
          board(bottom)(j) = 0
          bottom -= 1
        }
      })
      set.clear
    }
    board
  }
}

File: src/main/java/com/vc/medium/L723.java
package com.vc.medium;

class L723 {
    public int[][] candyCrush(int[][] board) {
        if(board == null || board.length == 0) return board;

        int n = board.length;
        int m = board[0].length;

        boolean found = true;
        while(found) {
            found = false;
            //Crush
            for(int row = 0; row < n; row++) {
                for(int col = 0; col < m; col++) {
                    //Already visited
                    if(board[row][col] == 0) continue;

                    int val = Math.abs(board[row][col]);

                    //check horizontal
                    if(col + 2 < m
                            && Math.abs(board[row][col + 1]) == val && Math.abs(board[row][col + 2]) == val) {
                        found = true;
                        int c = col;
                        while(c < m && Math.abs(board[row][c]) == val) {
                            board[row][c] = -1 * val;
                            c++;
                        }
                    }

                    //check vertical
                    if(row + 2 < n
                            && Math.abs(board[row + 1][col]) == val && Math.abs(board[row + 2][col]) == val) {
                        found = true;
                        int r = row;
                        while(r < n && Math.abs(board[r][col]) == val) {
                            board[r][col] = -1 * val;
                            r++;
                        }
                    }
                }
            }


            //System.out.println("Crush is finished!");

            //Fall
            if(found) {
                for(int col = 0; col < m; col++) {
                    int storeRow = n - 1;
                    for(int row = n - 1; row >= 0; row--){
                        if(board[row][col] > 0) {
                            board[storeRow--][col] = board[row][col];
                        }
                    }

                    while(storeRow >= 0) board[storeRow--][col] = 0;
                }
            }

            //System.out.println("Fall is finished!");

        }

        return board;
    }
}
